docs/src/adr/007adr-restructuring.md

```diff
@@ -0,0 +1,161 @@
+
+Any implementation of a L3 transformation must be explicitly annotated and 
+special care should be made during PR review to ascertain type-correctness. Whenever possible, unit tests should test for type correctness.
+
+# L4: Type analysis
+
+In L4, we have implementations of the following concepts:
+
+ - Any calculus in a TLA+ type system, including:
+    - Unification
+    - Sub/supertype relations 
+ - Any type-related static analysis, such as:
+    - Type checking
+    - Type inference
+
+# L5: Self-typing Builder:
+
+In L5, we have implementations of the following concepts:
+
+ - The TLA-IR builder
+
+# L6: Type-guaranteed IR manipulation
+
+In L6, we have implementations of the following concepts:
+
+ - TLA+ transformations that:
+    - do not introduce operators excluded from L0
+    - do not manually introduce IR constructors and type-tags, but use the self-typing builder instead
+
+## 3. Apalache core
+
+L7+ make up the Apalache core. The concepts of these levels are packaged as follows:
+
+ 1. `apa-base` : L7-L9
+ 2. `apa-pass` : L10
+ 3. `apa-tool` : L11+
+
+If individual passes in L10 are deemed complex enough, they may be placed in their own package (e.g. `apa-bmc`)
+
+# L7: Extensions of TLA+
+
+In L7, we have implementations of the following concepts:
+
+ - Any Apalache-specific operators, introduced into the IR (e.g. `ApalacheOper`)
+ - Extensions of any printer/reader/writer classes with support for the above specific operators
+
+# L8: Apalache-specific transformations
+
+In L8, we have implementations of the following concepts:
+
+ - TLA+ transformations that introduce operators from L7
+
+Like L3 transformations, if the builder is not used, the transformation should be explicitly marked and inspected for type-correctness.
+
+# L9: Glue
+
+In L9, we have implementations of the following concepts:
+
+ - Classes and traits used to define Apalache workflow or auxiliary functions, but not core functionality, such as:
+    - Exception handling
+    - Logging
+ - Classes and traits related to 3rd party technologies, such as:
+    - SMT
+    - TLC config
+
+# L10: Pass implementations
+
+In L10, we have implementations of the following concepts:
+
+ - Apalache passes and related infrastructure
+
+# L11: Module implementations
+
+In L11, we have implementations of the following concepts:
+
+- Apalache modules
+
+# L12: Tool wrapper:
+
+In L12, we have implementations of the following concepts:
+
+ - Apalache CMD interface
+
+# L13: GUI
+
+In L13, we have implementations of the following concepts:
+
+ - Graphical interface, if/when one exists
+
+## 4. Aliases, factories and exceptions
+
+Any alias or factory belongs to the lowest possible level required to define it. For example, `type uidToExMap = Map[UID, TlaEx]` belongs to L0, since `UID` and `TlaEx` are both L0 concepts, so it should be defined in an L0 alias package, even if it is only being used in a package of a higher level. 
```

Comment:
I'm not sure about this guide line.

The point of aliases and such is to abbreviate the names of something for use *in a particular context*. This principle would seem to make that intended use disallowed.

Reply:
Good point, I was thinking more of global aliases (e.g. `type TlaExTransformation = TlaEx => TlaEx`) when I wrote this bit, but you're right, it should be more clear w.r.t. local aliasing.